protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,131 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	"github.com/nats-io/nats.go"
+)
+
+// Protocol is a reference implementation for using the CloudEvents binding
+// integration. Protocol acts as both a NATS client and a NATS handler.
+type Protocol struct {
+
+	// nats connection
+	conn      *nats.Conn
+	url       string
+	natsOpts  []nats.Option
+	connOwned bool
+
+	receiver *receiver
+
+	sender *sender
+}
+
+// NewProtocol creates a new NATS protocol.
+// Use WithURL() OR WithConnection()
+// Use WithConsumerConfig() OR WithOrderedConsumerConfig()
+func NewProtocol(ctx context.Context, sendSubject string, opts ...ProtocolOption) (*Protocol, error) {
+	p := &Protocol{}
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if err := p.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs and we created the nats connection
+		if p.connOwned && p.conn != nil && errConnection != nil {
+			p.conn.Close()
+		}
+	}()
+
+	optsWithConnection := []ProtocolOption{}
+	optsWithConnection = append(optsWithConnection, opts...)
+
+	// if a URL was given create the nats connection
+	if p.conn == nil {
+		p.conn, errConnection = nats.Connect(p.url, p.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+		// This will add a WithConnection to the already existing WithURL in optsWithConnection.
+		optsWithConnection = append(optsWithConnection, WithConnection(p.conn))
+	}
+
+	if p.receiver, errConnection = NewReceiver(ctx, optsWithConnection...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	if p.sender, errConnection = NewSender(ctx, sendSubject, optsWithConnection...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	return p, nil
+}
+
+// Send implements Sender.Send
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) error {
+	return p.sender.Send(ctx, in, transformers...)
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	return p.receiver.OpenInbound(ctx)
+}
+
+// Receive implements Receiver.Receive
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	return p.receiver.Receive(ctx)
+}
+
+// Close implements Closer.Close
+func (p *Protocol) Close(ctx context.Context) error {
+	if p.connOwned {
```

Comment:
See my comment on this field. IMHO we can remove this field, document that a user has to call Close on the protocol, i.e., it's in their control when the connection will be closed, and therefore avoid duplicate calls to Close (is it already idempotent?) e.g., from your constructor where you also close the connection in a defer call when an error occurs.